Add specific cluster error info, shard info and additional metadata for CCS when minimizing roundtrips

quux00 opened the pull request:
For CCS searches with ccs_minimize_roundtrips=true, when an error is returned, it is unclear which cluster
caused the problem. This commit adds additional accounting and error information to the search response
for each cluster involved in a cross-cluster search.

The `_clusters` section of the SearchResponse has a new `details` section added with an entry for each cluster 
(remote and local). It includes status info, shard accounting counters and error information that are added 
incrementally as the search happens.

The search on each cluster can be in one of 5 states:
`RUNNING`
`SUCCESSFUL` - all shards were successfully searched (successful or skipped)
`PARTIAL` - some shard searches failed, but at least one succeeded and partial data has been returned
`SKIPPED` - no shards were successfully searched (all failed or cluster unavailable) when skip_unavailable=true
`FAILED` - no shards were successfully searched (all failed or cluster unavailable) when skip_unavailable=false

A new `SearchReponse.Cluster` object has been added. Each `TransportSearchAction.CCSActionListener`
(one for each cluster) has a reference to a separate Cluster instance and updates once it gets back information 
from its cluster.

The `SearchResponse.Clusters` object only uses the new `Cluster` object for CCS minimize_roundtrips=true. 
For local-only searches and CCS minimize_roundtrips=false, it uses the current (immutable) Clusters object as before.

quux00 committed:
First attempt.

Added SearchResponse.Cluster object. Modified MRT=true pathways to use a shard Clusters object
which has references to the per-cluster SearchResponse.Cluster object.

Tests checked and passing
1) basic happy path manual (no error pathways checked yet)
2) org.elasticsearch.action.search.TransportSearchActionTests
3) org.elasticsearch.search.CCSDuelIT.testMatchAll

Transport layer was modified (Writable), but XContent has not yet been modified.

quux00 committed:
Added 'details' of Cluster to XContent output

quux00 committed:
MP-README.md

quux00 committed:
MutableSearchResponse no longer needs special in updateFinalResponse for ccsMRT

TSA onFailure - adding to Cluster object for both skip_unavailable = true and false

quux00 committed:
Partially improved search-across-clusters.asciidoc. Leaving the rest until figure out exact response format

quux00 committed:
Removed boolean isFinalLocalReduce from MutableSearchResonse.updatePartialResponse

quux00 committed:
Removed remoteClusters count from SearchResponse.Clusters. All tests (my subset) pass.

quux00 committed:
Removed finalState setting in SearchResponse.Clusters and added asserts to SearchResponse.Clusters(StreamInput) ctor

Added logging of CCS errors in TransportSearchAction.

quux00 committed:
Updated TransportVersion.

Added hasPartialResults to SearchResponse.Clusters and isPartialResponse is MutableSearchResponse that calls it
Added production logging of errors in TSA for CCS failures.

quux00 committed:
Changed "success" Cluster status to "successful".

Added logCCSError to TransportSearchAction.

quux00 committed:
Added test to CrossClusterAsyncSearchIT that tests _cluster/details for successful CCS

quux00 committed:
Updated CrossClusterAsyncSearchIT to have multiple tests of the the new Clusters/Cluster objects

Modified AbstractMultiClustersTestCase to have a way to set skip_unavailable setting for remote cluster config.

javanna commented:
> 1. Should we add a `failed` counter to the `_clusters` output? If not, do we account for them in the `_clusters/skipped` counter?

Sounds like a good idea to me, although skipped and failed are never going to be greater than 0 at the same time? Should we hide either of the two depending on skip_unavailable? Or maybe that's too magic?

> 2. How do we want to label the local cluster in the `_clusters/details` section? In the ES codebase it just has the alias "" (empty string), but that is potentially confusing to end users, so I've put `(local)` as the proposed identifier. This also matches what I did for the profile output of a search here: https://github.com/elastic/elasticsearch/pull/96396

sounds good to me.

> 4. Should we include the `search_duration` field?

I would call it `took`

> 5. Should we include the `percent_shards_successful` field?

I would not include it, it can easily be computed.

> 6. Should we include the `_cluster/details` section by default? With say 100 clusters, this section is going to be quite long. Each entry in the `details` section is currently 8 lines of JSON when successful and another 8 to several dozen lines if there a lot of shard failures to specify the errors for. Should it be put behind a flag like `detailed=true` ([like the Task API has](https://www.elastic.co/guide/en/elasticsearch/reference/current/tasks.html#tasks-api-query-params)).

I guess it won't matter much if Kibana ends up always providing the flag? What is the plan there in terms of consuming the additional output?

> 7. Should we have a more limited `_cluster/details` output by default and only put the full output behind a `detailed=true` flag? If yes, what information should be displayed by default?

I am not sure we would be able to trim it down considerably. Maybe we could remove the shards information from it.  I am not entirely sure, and depends again on how we expect this info to be consumed.

> 8. For CCS minimize_roundtrips=true, should we remove or zero out the `_shards` section since (a) it is somewhat redundant to the information supplied in the _clusters section when the search is successful, (b) when the search is running it always inaccurately displays only the local shard count for `total` shards, and most importantly, (c) that section is inaccurate at completion when a shard is skipped or failed because it cannot be reached. In that case we don't have any information on how many shards the unreachable/failing remote cluster has.  (For context - the _shards section is what the CCS MRT=false setting uses for it's status and errors.)

I suspect this applies to async search as it exposes partial state but less to _search. I would not consider this necessary for now.

quux00 commented:
> Sounds like a good idea to me, although skipped and failed are never going to be greater than 0 at the same time? Should we hide either of the two depending on skip_unavailable? Or maybe that's too magic?

Yes, you can have both `skipped` and `failed` with a count > 0. Example scenario: 3 remote clusters in the search. One completes successfully, the other two completely fail (all shards fail or cluster not available), where one is `skip_unavailable=true` and the other is `skip_unavailable=false`. The counters would then be:
```
      "total": 3,
      "successful": 1,
      "skipped": 1,
      "failed": 1
```

quux00 committed:
PR feedback.

Changed searchDuration to took.
Updated TransportVersion
Removed percent_shards_successful
Moved per cluster shard accounting into a new _shards sub field.
Fixed failing CrossClusterSearchIT by adjusting the setup code in AbstractMultiClustersTestCase

quux00 committed:
Changed errors to failures in details section of clusters

quux00 committed:
Cluster object now takes indexExpression and the _clusters/details entry now has "cluster:index" notation
using the original indexExpression (not resolved concrete indices).

quux00 committed:
Made SearchResponse.Cluster instance fields thread safe (to the limited degree needed)

quux00 committed:
Marking each Cluster with timedOut and PARTIAL status if its SearchResponse times out

quux00 committed:
Moved indexExpression out of details/cluster-name section and into own field

It was too hard to test with docs yaml test in that position.
Updated search-across-clusters.asciidoc and tests are passing.

quux00 committed:
Created common static method ccsClusterInfoUpdate in TransportSearchAction to reduce code duplication

quux00 committed:
Cleaned up search-across-clusters.asciidoc

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

quux00 committed:
clean up

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 committed:
Added 2 XContent tests for new Clusters object format to AsyncSearchResponseTests

quux00 committed:
randomSearchResponse(boolean ccs) added used by AsyncSearchResponseTests

quux00 committed:
randomSearchResponse(boolean ccs) added used by AsyncStatusResponseTests

quux00 committed:
AsyncStatusResponseTests.createTestInstance uses AsyncSearchResponseTests.createCCSClusterObject

quux00 committed:
final test changes for AsyncStatusResponseTests

quux00 commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -507,53 +518,99 @@ public Clusters(int total, int successful, int skipped, int remoteClusters, bool
          */
-        public Clusters(int total, int successful, int skipped) {
-            this(total, successful, skipped, true);
+        public Clusters(
```
There are two separate ways to use `SearchResponse.Clusters`.

The original way is that `SearchResponse.Clusters` is immutable and just holds counters for total, successful, and skipped. A typical usage is just to use SearchResponse.Clusters.EMPTY when you don't even want to track that accounting (local only searches). For CCS MRT=false, an immutable SearchResponse.Clusters is created after initially contacting each remote cluster (in order to do can-match via SearchShards API) and the Clusters object is never updated after that (even if one of the remote clusters later fails during the search).

For CCS MRT=true in this PR, the Clusters object is now composed of several underlying `SearchResponse.Cluster` object, each of which is updated separately as the remote clusters report back. So in this model an initial Clusters object is created at the beginning of the search and it is updated periodically from separate callback threads.

This model can also be used in a future PR for CCS MRT=false, but that is out of scope for this PR.

quux00 commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -604,28 +676,263 @@ public boolean equals(Object o) {
+        private final AtomicInteger skippedShards;
+        private final AtomicInteger failedShards;
+        private final AtomicLong took;  // search latency in millis for this cluster sub-search
+        private final AtomicBoolean timedOut;
```
The thread safety model of this class does not use synchronization because it holds to a single-writer principle elsewhere in the code.

The Cluster object is updated by one ActionListener callback thread. No other thread should ever update a given Cluster object. (Each is updated by a separate single ActionListener). 

However, the Cluster objects do need to be read by queries (e.g., `GET _async_search/:id`), so all variables are set as volatile or Atomic (or final for `clusterAlias` and `indexExpression`) to allow getting latest state for status queries.

Note: the status access (e.g., `GET _async_search/:id`) is "best effort" since there is no synchronization. It is possible that while the status is being read for a status response, the Cluster is being actively updated by a callback thread. This is considered acceptable since that would only happen for "intermediate state". The final state when the search is completed will always be fully accurate.

quux00 replied:
Also I think technically `took` and `timedOut` could just be volatile primitives, since they are always just set once and never incremented.

javanna replied:
Is it an option to create a new Cluster object at every change and keep the class immutable? I think that would be easier to reason about compared to all the mutable state it has now.

quux00 replied:
For MRT=true, yes, because each cluster has only a single callback thread, so you have a single writer per Cluster.

But for MRT=false, the callback thread is at the granularity of a shard. So during the CanMatch (SearchShards) and during the Query/Fetch Phase you have N threads (N=number of shards per cluster) all potentially updating the same Cluster object with specific shard info. So either you need a shard thread safe Cluster object OR, if we want to use immutable objects, we need a shared per-Cluster lock so that each shard callback thread and safely increment the successful/skipped/failed counters and the failure list without a data race that loses information.

I think both are feasible. For the latter, I think we would need the shared Clusters object (which no longer can be immutable for CCS) to construct a set of lock objects, one per Cluster. Then the shard callback threads (MRT=false) would need to obtain that lock in order to create a new Cluster object with the updated information and then overwrite the existing and then release the lock.

If build such as locking system, we might want to then wire the MRT=true paths to use it as well, though there with a single writer per cluster it isn't necessary.

Let me know your view (or if there's another option I'm not thinking of) on how to proceed.

quux00 replied:
In the latest push I changed the `SearchResponse.Cluster` to be fully immutable. The `SearchResponse.Clusters` object holds an unmodifiable Map<String, AtomicReference<Cluster>>, where the key is the clusterAlias.

For MRT=true (this PR), each CCSActionListener (one per cluster) gets a reference to the AtomicReference holding the Cluster it is searching. It will then do a CAS operation to update state with a new Cluster object. This isn't strictly necessary for the MRT=true paths, since there is a single writer to the Map per Cluster.

But it sets up the concurrency model for MRT=false, where the updates will be done per shard, not per Cluster. For that work (future ticket), each updater thread will update a Cluster object by retrieving the existing Cluster object, incrementing the appropriate counters and other state (e.g., status or timedOut or failure lists) and then doing a CAS back into the AtomicReference. If that fails, it will grab the new Cluster being held by the AtomicReference and re-attempt to increment/add data to a new Cluster and again try to CAS it into the AtomicReference.

quux00 commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -1328,6 +1409,30 @@ private void maybeFinish() {
+     * @param clusterAlias cluster on which the failure occurred
+     * @param skipUnavailable the skip_unavailable setting of the cluster with the search error
+     */
+    private static void logCCSError(ShardSearchFailure f, String clusterAlias, boolean skipUnavailable) {
```
This "feature" was requested by Product at a recent meeting - to start logging any CCS errors that occur so that we can begin to discover what types of CCS issues happen in large production environments, especially as we don't yet have any CCS telemetry in place.

quux00 replied:
OK, I can do that.

quux00 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncStatusResponse.java:
```diff
@@ -189,7 +189,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         RestActions.buildBroadcastShardsHeader(builder, params, totalShards, successfulShards, skippedShards, failedShards, null);
         if (clusters != null) {
-            builder = clusters.toXContent(builder, null);
+            builder = clusters.toXContent(builder, params);
```
Bug fix that was uncovered by the newly added code in this PR.

quux00 commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -604,28 +676,263 @@ public boolean equals(Object o) {
         @Override
         public int hashCode() {
+            /// MP TODO: not sure what to do about successful and skipped here, since they are calculated differently
+            /// MP TODO: between CCS MRT=true and other search types and are not immutable for CCS
```
I need help here. I'm not sure what to do with these, since the Clusters object is no longer immutable under the CCS MRT=true scenario. I'm not clear when equals and hashCode is called for Clusters.

quux00 replied:
Thanks. I removed the comments (next push).

quux00 commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -570,29 +627,44 @@ public int getTotal() {
          */
         public int getSuccessful() {
-            return successful;
+            if (clusterInfo.isEmpty()) {
```
`clusterInfo` is always empty for local-only searches and CCS MRT=false, so it can be used as the proxy for when Clusters is immutable vs. updatable.

quux00 replied:
I have an idea for refactoring. Will be in the next push.

quux00 commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -570,29 +627,44 @@ public int getTotal() {
+                    .stream()
+                    .filter(c -> c.getStatus() == Cluster.Status.SUCCESSFUL || c.getStatus() == Cluster.Status.PARTIAL)
+                    .count();
+            }
```
As mentioned above, from a thread-safety and accuracy point of view, this is best effort, not guaranteed accurate as the Cluster objects might in the act of being updated while a "status" update query is reading the Cluster object state. For performance reasons, this is considered a fair tradeoff, since the final status is the only one guaranteed to be truly accurate.

quux00 committed:
Removed usage of Deprecated Clusters constructor from SearchResponseTests

quux00 committed:
CrossClusterAsyncSearch - ensuring that tests that do not use the SearchListenerPlugin negate it (shut it off) before running tests

quux00 commented:
**Example output from manual tests**

<details><summary>(toggle) Skipped cluster due to failures on skip_unavailable=true</summary>

Query:

In this case the `web_traffic` index does not exist on the `remote2` cluster.

```
POST *,remote*:blogs,remote2:web_traffic,blogs/_async_search?ccs_minimize_roundtrips=true
{
  "size": 1,
  "query": {
    "function_score": {
      "query": {
        "wildcard": {
          "url": "/*"
        }
      },
      "script_score": {
        "script": "int m = 13;\n  double u = 1.0;\n  double z = 2.0;\n for (int x = 0; x < m; ++x) {\n for (int y = 0; y < 10000; ++y) {\n u=Math.log(y);\n  z=z + (Math.log(x) - Math.log(y));\n }\n }\n return u;\n"
      }
    }
  },
  "aggs": {
    "indexgroup": {
      "terms": {
        "field": "_index"
      }
    }
  }
}
```

Response:

```
{
  "id": "FnE3TjdSbnAtVGx1ekxWYVdEQTczbWcaRTAzaTNvV2xScnFNRklDUWpKWFdEdzo5MzA=",
  "is_partial": true,
  "is_running": false,
  "start_time_in_millis": 1690819892954,
  "expiration_time_in_millis": 1691251892954,
  "completion_time_in_millis": 1690819905011,
  "response": {
    "took": 12057,
    "timed_out": false,
    "num_reduce_phases": 3,
    "_shards": {
      "total": 6,
      "successful": 6,
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 2,
      "skipped": 1,
      "details": {
        "(local)": {
          "status": "successful",
          "indices": "*,blogs",
          "took": 12050,
          "timed_out": false,
          "_shards": {
            "total": 3,
            "successful": 3,
            "skipped": 0,
            "failed": 0
          }
        },
        "remote2": {
          "status": "skipped",
          "indices": "blogs,web_traffic",
          "timed_out": false,
          "failures": [
            {
              "shard": -1,
              "index": null,
              "reason": {
                "type": "index_not_found_exception",
                "reason": "no such index [web_traffic]",
                "index_uuid": "_na_",
                "resource.type": "index_or_alias",
                "resource.id": "web_traffic",
                "index": "web_traffic"
              }
            }
          ]
        },
        "remote1": {
          "status": "successful",
          "indices": "blogs",
          "took": 11755,
          "timed_out": false,
          "_shards": {
            "total": 3,
            "successful": 3,
            "skipped": 0,
            "failed": 0
          }
        }
      }
    },
    "hits": {
      "total": {
        "value": 2798,
        "relation": "eq"
      },
      "max_score": 9.21024,
      "hits": [
        { ... }
      ]
    },
    "aggregations": {
      "indexgroup": {
        "doc_count_error_upper_bound": 0,
        "sum_other_doc_count": 0,
        "buckets": [
          {
            "key": "blogs",
            "doc_count": 1399
          },
          {
            "key": "remote1:blogs",
            "doc_count": 1399
          }
        ]
      }
    }
  }
}
```

</details>

quux00 committed:
Added skipUnavailable to ccsClusterInfoUpdate. Minor docs test bug fix

quux00 commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -495,6 +505,7 @@ public static class Clusters implements ToXContentFragment, Writeable {
          * @param remoteClusters number of remote clusters in the search
          * @param ccsMinimizeRoundtrips specifies whether a CCS search is using minimizeRoundtrips feature
          */
+        @Deprecated  // for CCS use the Clusters(OriginalIndices, Map<String, OriginalIndices>, boolean) constructor
```
Nothing in the ES code is using this constructor any longer. I'd like to remove it, but since it was marked as public and released in 8.8, I'm assuming I cannot remove it in case any 3rd party modules/plugins are using it?

javanna commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -714,25 +744,67 @@ private static ActionListener<SearchResponse> createCCSListener(
+                ccsClusterInfoUpdate(searchResponse, cluster, skipUnavailable);
+                Instant now = Instant.now();
+                Instant start = Instant.ofEpochMilli(startTime);
+                cluster.setTook(Duration.between(start, now).toMillis());
```
can you remind me why we need to compute the took time here? is it not included in the search response obtained from the remote cluster?

javanna reviewed (commented):
left a couple comments, looks headed in the right direction!

quux00 committed:
PR feedback.

Removed deprecated Clusters constructor.

Created common determineCountFromClusterInfo method for dynamic counts
in Clusters from underlying Cluster objects.

Took on Cluster objects is now set from the underlying single cluster SearchResponse.

quux00 committed:
Intermediate commit of making SearchResponse.Cluster immutable before changing to AtomicReference - not working/incomplete

quux00 committed:
Changed SearchResponse.Cluster objects to be immutable.

The SearchResponse.Clusters object now uses an immutable HashMap<String, AtomicReference<Cluster>>
rather than ConcurrentHashMap<String, Cluster>, since the concurrency control is
now a CAS operation on the underlying AtomicReference<Cluster>> not an overwrite into
the HashMap.

Changed took in SearchResponse.Cluster to be a TimeValue rather than long to match
the usage in SearchResponse.

quux00 committed:
Minor cleanup and javadoc improvements

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you. *Note* that since this PR is labelled `release highlight`, you need to update the changelog YAML to fill out the extended information sections.

javanna commented on docs/reference/search/search-your-data/search-across-clusters.asciidoc:
```diff
@@ -208,15 +230,53 @@ The API returns the following response:
+    "skipped": 0,
+    "details": {
+      "(local)": {            <1>
+        "status": "successful",
```
quick q: there is no way to represent this as http status code? I was looking at this and being surprised with some new type of info being returned and possibly not knowing what the options are and how to interpret it. it is though quite self-explanatory. Do we have other places where we return similar information in other API?

quux00 replied:
HTTP status code wouldn't work since PARTIAL and SUCCESSFUL are both "successful" (200 OK) for clusters (similar to how for shards "skipped" and "successful" are "success" statuses). And what HTTP status would we use for "skipped"? I guess we could use 400, but the overall query is still a success (200 OK). I think that would fairly cryptic to end users.

The status value here is meant to refer to the cluster "accounting" options above: total/successful/skipped. I plan to add "failed" to the accounting list at some point, but left it out of this PR since it is getting large.

So to address the concern of "I don't know what all possible states are", we could list those in the counter section:
```
  "_clusters": {
    "total": 5,
    "successful": 2,
    "partial": 1,    // not present
    "skipped": 1,
    "failed": 0,     // plan to add later
    "details": {
      "(local)": {           
        "status": "successful",
         ...
       }
      "remote1": {           
        "status": "skipped",
```

javanna replied:
Thanks for expanding. The proposed change makes sense to me, as a lower priority follow-up.

javanna commented on docs/reference/search/search-your-data/search-across-clusters.asciidoc:
```diff
@@ -360,20 +429,22 @@ The API returns the following response:
-// TESTRESPONSE[s/"hits": \[\]/"hits": $body.response.hits.hits/]
+// TEST[skip: terminated_early is absent from final results so is hard to reproduce here]
+
+x // TESTRESPONSE[s/FklQYndoTDJ2VEFlMEVBTzFJMGhJVFEaLVlKYndBWWZSMUdicUc4WVlEaFl4ZzoxNTU=/$body.id/]
```
what are the `x` s at the beginning of each line?

quux00 replied:
Oh, sorry. Left over during edits. Good catch. I'll remove it.

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -112,20 +105,6 @@ synchronized void updatePartialResponse(
-            );
-            this.clusters = newClusters;
-            logger.debug("Updating Clusters info to indicate that the local cluster search has completed: {}", newClusters);
-        }
```
Nice, no need to manually do this as each cluster is going to update its state automatically>

javanna reviewed (approved):
Left some additional questions. No blockers, all minor things. Feel free to merge when ready.

quux00 commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -614,18 +682,246 @@ public int hashCode() {
+                    if (failedShards != null) {
+                        builder.field("failed", failedShards);
+                    }
+                    builder.endObject();
```
I'm not sure what you are asking/proposing. Do you mean that I should move this "toXContent" functionality up to the `Clusters.toXContent`? 

That's where this one is called from. The Clusters.toXContent calls each individual `Cluster.toXContent`:

```
        @Override
        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
            if (total > 0) {
                builder.startObject(_CLUSTERS_FIELD.getPreferredName());
                builder.field(TOTAL_FIELD.getPreferredName(), total);
                builder.field(SUCCESSFUL_FIELD.getPreferredName(), getSuccessful());
                builder.field(SKIPPED_FIELD.getPreferredName(), getSkipped());
                // TODO: add FAILED_FIELD
                if (clusterInfo.size() > 0) {
                    builder.startObject("details");
                    for (AtomicReference<Cluster> cluster : clusterInfo.values()) {
                        cluster.get().toXContent(builder, params);   // call to each Cluster.toXContent here
                    }
                    builder.endObject();
                }
                builder.endObject();
            }
            return builder;
        }
```

The model I used here for the REST layer "serialization" is patterned after how we handle the Transport layer serialization, where `Clusters.writeTo` (and read) relies on each Cluster to do its portion of the Writeable action. Here is the `Clusters.writeTo` method:

```
        @Override
        public void writeTo(StreamOutput out) throws IOException {
            out.writeVInt(total);
            out.writeVInt(successful);
            out.writeVInt(skipped);
            if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_500_052)) {
                if (clusterInfo != null) {
                    List<Cluster> clusterList = clusterInfo.values().stream().map(AtomicReference::get).toList();
                    out.writeList(clusterList);
                } else {
                    out.writeList(Collections.emptyList());
                }
            }
        }
```

Sorry if I missed what you are suggesting. Let me know what you'd prefer.

javanna replied:
it's fine and not so important. I was referring to the fact that search already outputs a _shards section that looks the same as this one, and I wondered if those two toXcontent could be unified. It won't save a lot of code, but it would make it clear that the two are aligned in terms of naming for a good reason. This could totally be a follow-up.

quux00 replied:
Oh, I see. I can try that out in the next ticket (I have two or three follow-ons to this one).

quux00 commented on server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:
```diff
@@ -720,25 +742,109 @@ private static ActionListener<SearchResponse> createCCSListener(
+            String indexExpression = orig.getIndexExpression();
+            SearchResponse.Cluster updated = new SearchResponse.Cluster(clusterAlias, indexExpression, status, failures);
+            swapped = clusterRef.compareAndSet(orig, updated);
+        } while (swapped == false);
```
This concurrency model was primarily designed with MRT=false in mind but I wanted one model for both systems so I implemented it here as well. 

However, in my next PR (CCS fail fast), this model is also now useful for MRT=true, because other Cluster objects will have their status updated to `CANCELLED` (a new status not in this PR) in the callback thread of the Cluster that is failing the search, so you could then have one thread trying to update the Cluster object to have, say, `SUCCESSFUL` status, and the "cancellation" thread updating that same Cluster reference to `CANCELLED` status at the same time, so this CAS concurrency control model will be useful to 1) avoid losing state added by another thread and 2) allow a sticky state machine (once the Cluster is moved to CANCELLED it should not be changed).

quux00 committed:
PR feedback

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you. *Note* that since this PR is labelled `release highlight`, you need to update the changelog YAML to fill out the extended information sections.

quux00 commented:
@elasticsearchmachine generate changelog
